lib/pure/complex.nim

```diff
@@ -9,78 +9,140 @@
 
 
 
-## This module implements complex numbers.
-{.push checks:off, line_dir:off, stack_trace:off, debugger:off.}
-# the user does not want to trace a part
-# of the standard library!
 
 
-import
-  math
+## This module implements complex numbers.
+## Complex numbers are currently implemented as generic on a 64-bit or 32-bit float.
+
+{.push checks: off, line_dir: off, stack_trace: off, debugger: off.}
+# the user does not want to trace a part of the standard library!
 
-const
-  EPS = 1.0e-7 ## Epsilon used for float comparisons.
+import math
 
 type
-  Complex* = tuple[re, im: float]
-    ## a complex number, consisting of a real and an imaginary part
+  Complex*[T: SomeFloat] = object
+    re, im: T
+    ## A complex number, consisting of a real and an imaginary part.
+  Complex64* = Complex[float64]
+    ## Alias for a pair of 64-bit floats.
+  Complex32* = Complex[float32]
+    ## Alias for a pair of 32-bit floats.
+
+proc complex*[T: SomeFloat](re: T; im: T = 0.0): Complex[T] =
+  result.re = re
+  result.im = im
+
+proc complex32*(re: float32; im: float32 = 0.0): Complex[float32] =
+  result.re = re
+  result.im = im
+
+proc complex64*(re: float64; im: float64 = 0.0): Complex[float64] =
+  result.re = re
+  result.im = im
+
+template im*(arg: typedesc[float32]): Complex32 = complex[float32](0, 1)
+template im*(arg: typedesc[float64]): Complex64 = complex[float64](0, 1)
+template im*(arg : float32): Complex32 = complex[float32](0, arg)
+template im*(arg : float64): Complex64 = complex[float64](0, arg)
+
+proc toComplex*[T: SomeNumber](x: T): auto =
+  ## Convert ``x`` to Complex.
+  when T is SomeInteger:
+    result.re = float(x)
```

Comment:
This is too arbitrary, `toComplex` should not convert to Complex128 by default for integers.

As I said here https://github.com/nim-lang/Nim/pull/9452#discussion_r226830525, let the user define the return typedesc they want.

> `toComplex` should take the result type as a typedesc or generic parameters. We don't care about the input type. This makes `toComplex64` and `toComplex128` unnecessary and avoid the awkward implementation in toComplex

Reply:
You are right. I actually wanted to remove this proc entirely, but I forgot to do so.